Fix the race condition in the test pipeline for filtered events.

mashhurs opened the pull request:
### Description
We observed a weird behaviour with [date-filter CI](https://app.travis-ci.com/github/logstash-plugins/logstash-filter-date/jobs/636000806?serverType=git) that shows `filtered_events` (you can see the error belo) in the test pipeline looks like returning an empty array. I have investigated filter-date source (below) that there is no possibility of being array an empty.
So, it seems to me a race condition with `filtered_events` in test pipeline.

```
// date-filter receive method which is called from multi_receive
public List<RubyEvent> receive(List<RubyEvent> rubyEvents) {
    for (RubyEvent rubyEvent : rubyEvents) {
      Event event = rubyEvent.getEvent();

      switch (executeParsers(event)) { // handles exception cases and returns proper value
        case FIELD_VALUE_IS_NULL_OR_FIELD_NOT_PRESENT:
        case IGNORED:
          continue;
        case SUCCESS:
          if (successHandler != null) {
            successHandler.handle(rubyEvent);
          }
          break;
        case FAIL: // fall through
        default:
          for (String t : tagOnFailure) {
            event.tag(t);
          }
          if (failureHandler != null) {
            failureHandler.handle(rubyEvent);
          }
      }
    }
    return rubyEvents;
  }

```

```
[2025-10-18T17:03:19,223][ERROR][logstash.testpipeline    ] Pipeline worker error, the pipeline will be stopped {:pipeline_id=>"main", :exception=>"Java::JavaLang::ArrayIndexOutOfBoundsException", :error=>"Index 0 out of bounds for length 0", :stacktrace=>"org.jruby.dist/org.jruby.specialized.RubyArrayOneObject.copyInto(RubyArrayOneObject.java:103)\norg.jruby.dist/org.jruby.RubyArray.splice(RubyArray.java:1090)\norg.jruby.dist/org.jruby.RubyArray.concat(RubyArray.java:1823)\norg.jruby.dist/org.jruby.RubyArray.concat(RubyArray.java:1818)\nusr.share.logstash.vendor.jruby.lib.ruby.gems.shared.gems.logstash_minus_devutils_minus_2_dot_6_dot_2_minus_java.lib.logstash.test_pipeline.RUBY$method$filtered_events$0(/usr/share/logstash/vendor/jruby/lib/ruby/gems/shared/gems/logstash-devutils-2.6.2-java/lib/logstash/test_pipeline.rb:123)\nusr.share.logstash.vendor.jruby.lib.ruby.gems.shared.gems.logstash_minus_devutils_minus_2_dot_6_dot_2_minus_java.lib.logstash.test_pipeline.RUBY$block$events$1(/usr/share/logstash/vendor/jruby/lib/ruby/gems/shared/gems/logstash-devutils-2.6.2-java/lib/logstash/test_pipeline.rb:156)\norg.jruby.dist/org.jruby.runtime.CompiledIRBlockBody.yieldDirect(CompiledIRBlockBody.java:151)\norg.jruby.dist/org.jruby.runtime.BlockBody.yield(BlockBody.java:106)\norg.jruby.dist/org.jruby.runtime.Block.yield(Block.java:189)\norg.jruby.dist/org.jruby.ir.runtime.IRRuntimeHelpers.yield(IRRuntimeHelpers.java:494)\norg.jruby.dist/org.jruby.ir.targets.indy.YieldSite.yield(YieldSite.java:117)\nuri_3a_classloader_3a_.jruby.kernel.kernel.RUBY$method$tap$0(uri:classloader:/jruby/kernel/kernel.rb:19)\nusr.share.logstash.vendor.jruby.lib.ruby.gems.shared.gems.logstash_minus_devutils_minus_2_dot_6_dot_2_minus_java.lib.logstash.test_pipeline.RUBY$method$events$0(/usr/share/logstash/vendor/jruby/lib/ruby/gems/shared/gems/logstash-devutils-2.6.2-java/lib/logstash/test_pipeline.rb:154)\nusr.share.logstash.vendor.jruby.lib.ruby.gems.shared.gems.logstash_minus_devutils_minus_2_dot_6_dot_2_minus_java.lib.logstash.test_pipeline.RUBY$method$events$0$__VARARGS__(/usr/share/logstash/vendor/jruby/lib/ruby/gems/shared/gems/logstash-devutils-2.6.2-java/lib/logstash/test_pipeline.rb:153)\norg.jruby.dist/org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:139)\norg.jruby.dist/org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:112)\norg.jruby.gen.LogStash$$TestPipeline$$QueueBatchDelegator_62069443.events(org/jruby/gen/LogStash$$TestPipeline$$QueueBatchDelegator_62069443.gen:13)\norg.logstash.config.ir.CompiledPipeline$CompiledUnorderedExecution.compute(CompiledPipeline.java:358)\norg.logstash.execution.ObservedExecution.lambda$compute$0(ObservedExecution.java:17)\norg.logstash.execution.WorkerObserver.lambda$observeExecutionComputation$0(WorkerObserver.java:39)\norg.logstash.instrument.metrics.timer.NullTimerMetric.time(NullTimerMetric.java:17)\norg.logstash.execution.WorkerObserver.lambda$executeWithTimers$1(WorkerObserver.java:50)\norg.logstash.instrument.metrics.timer.NullTimerMetric.time(NullTimerMetric.java:17)\norg.logstash.execution.WorkerObserver.executeWithTimers(WorkerObserver.java:50)\norg.logstash.execution.WorkerObserver.observeExecutionComputation(WorkerObserver.java:38)\norg.logstash.execution.ObservedExecution.compute(ObservedExecution.java:17)\norg.logstash.execution.WorkerLoop.abortableCompute(WorkerLoop.java:113)\norg.logstash.execution.WorkerLoop.run(WorkerLoop.java:86)\nusr.share.logstash.logstash_minus_core.lib.logstash.java_pipeline.RUBY$block$start_workers$5(/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:317)\norg.jruby.dist/org.jruby.runtime.CompiledIRBlockBody.callDirect(CompiledIRBlockBody.java:141)\norg.jruby.dist/org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:64)\norg.jruby.dist/org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:58)\norg.jruby.dist/org.jruby.runtime.Block.call(Block.java:144)\norg.jruby.dist/org.jruby.RubyProc.call(RubyProc.java:354)\norg.jruby.dist/org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:111)\njava.base/java.lang.Thread.run(Thread.java:1583)", :thread=>"#<Thread:0x4afdb05e@[main]-pipeline-manager sleep>"}

```

mashhurs committed:
Fix the race condition in the test pipeline for filtered events.

mashhurs committed:
Version bump and add changelog.

donoghuc commented on lib/logstash/test_pipeline.rb:
```diff
@@ -61,12 +61,11 @@ class EventTrackingQueueReadClientDelegator
         # NOTE: can not use LogStash::MemoryReadClient#read_batch due its JavaObject wrapping
         @delegate = delegate.to_java(QueueReadClient)
         @processed_events = []
+        @mutex = Mutex.new
```
The stack trace does look like a race condition (likely between reset_events and filtered_events). Instead of managing the mutex ourselves i think the `Concurrent::Array` https://github.com/elastic/logstash/blob/915b69fee3f4608614349f96ea5242de908dc3e6/x-pack/spec/monitoring/inputs/metrics_spec.rb#L43 we use elsewhere would be a nice thread safe alternative. 
```ruby
# Use the threadsafe array
@processed_events = Concurrent::Array.new
# get
@processed_events.to_a
# filtered
@processed_events.concat(events)
# reset
@processed_events.clear
```

mashhurs replied:
`Concurrent::Array` is a good suggestion, thanks!
Yeah, it is hard to prove whether it is race condition but there is something happening when using `sample` (`sample({"mydate" => input})...`) method and `subject.get()` in its block.
With this change, we make sure `filtered_events` (appeared in the error) safely returns the `processed_events` where filter-date is trying to get the first element (raising an exception sometims now since it is an empty or nil).

mashhurs replied:
That's being said, I have changed from `Mutex` to `Concurrent:Array`.

mashhurs committed:
Use Concurrent::Array for the thread safety instead of manually managing with Mutex.

donoghuc reviewed (approved):
Looks good to me, though it looks like there are no automated tests for this repo? Were you able to do a manual test somewhere with this change?

mashhurs commented:
> Looks good to me, though it looks like there are no automated tests for this repo? Were you able to do a manual test somewhere with this change?

Yeah, testing was a bit tricky.
What I did was, moved the logstash-devutils repo under the logstash-filter-date and added `gem 'logstash-devutils', :path => "/usr/share/plugins/plugin/logstash-devutils"` to its Gemfile.
In case I have placed `RUN gem list` to make sure we didn't install the devutils-v2.6.2 - confirmed.

When running `./.ci-docker-setup.sh && ./.ci/docker-run.sh` I can see that that logstash-filter-date specs are using the updated version. To make 100% sure, I added just `puts` messages in the updated methods and here is the full log:

```
ls-plugins/logstash-filter-date (upstream-main*) » ./.ci/docker-run.sh                                                                                                      
++ dirname ./.ci/docker-run.sh
+ CURRENT_DIR=./.ci
+ cd .ci
+ export BUILDKIT_PROGRESS=plain
+ BUILDKIT_PROGRESS=plain
+ docker compose up --exit-code-from logstash
[+] Running 2/2
 ✔ Network ci_default       Created                                                                                                                                                                  0.0s 
 ✔ Container ci-logstash-1  Created                                                                                                                                                                  0.2s 
Attaching to logstash-1
logstash-1  | HOSTNAME=f3bc33c11c9e
logstash-1  | ELASTIC_CONTAINER=true
logstash-1  | LOGSTASH_PATH=/usr/share/logstash
logstash-1  | PWD=/usr/share/plugins/plugin
logstash-1  | ELASTICSEARCH_TREEISH=
logstash-1  | container=oci
logstash-1  | HOME=/usr/share/logstash
logstash-1  | LANG=C.UTF-8
logstash-1  | ELASTIC_STACK_VERSION=9.1.6
logstash-1  | DISTRIBUTION=
logstash-1  | TERM=xterm
logstash-1  | LOGSTASH_SOURCE=1
logstash-1  | INTEGRATION=
logstash-1  | LS_JAVA_OPTS=-Dls.cgroup.cpuacct.path.override=/ -Dls.cgroup.cpu.path.override=/ 
logstash-1  | SHLVL=1
logstash-1  | SNAPSHOT=false
logstash-1  | JRUBY_OPTS=-J--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED -J--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED -J--add-exports=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED -J--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED -J--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED -J--add-opens=java.base/java.security=ALL-UNNAMED -J--add-opens=java.base/java.io=ALL-UNNAMED -J--add-opens=java.base/java.nio.channels=ALL-UNNAMED -J--add-opens=java.base/sun.nio.ch=ALL-UNNAMED -J--add-opens=java.management/sun.management=ALL-UNNAMED -Xregexp.interruptible=true -Xcompile.invokedynamic=true -Xjit.threshold=0 -J-XX:+PrintCommandLineFlags -v -W1
logstash-1  | LC_ALL=C.UTF-8
logstash-1  | SPEC_OPTS="-fd"
logstash-1  | PATH=/usr/share/logstash/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/share/logstash/vendor/jruby/bin:/usr/share/logstash/jdk/bin
logstash-1  | _=/usr/bin/env
logstash-1  | + bundle exec rake test
logstash-1  | -XX:+AutoCreateSharedArchive -XX:ConcGCThreads=3 -XX:G1ConcRefinementThreads=10 -XX:GCDrainStackTargetSize=64 -XX:InitialHeapSize=195769920 -XX:MarkStackSize=4194304 -XX:MaxHeapSize=3132318720 -XX:MinHeapSize=6815736 -XX:+PrintCommandLineFlags -XX:ReservedCodeCacheSize=251658240 -XX:+SegmentedCodeCache -XX:SharedArchiveFile=/usr/share/logstash/vendor/jruby/lib/jruby-java21.0.8.jsa -XX:ThreadStackSize=2048 -XX:+UseCompressedOops -XX:+UseG1GC 
logstash-1  | jruby 9.4.13.0 (3.1.4) 2025-06-10 9938a3461f OpenJDK 64-Bit Server VM 21.0.8+9-LTS on 21.0.8+9-LTS +indy +jit [aarch64-linux]
logstash-1  | /usr/share/plugins/plugin/gradlew --no-daemon test
logstash-1  | To honour the JVM settings for this build a single-use Daemon process will be forked. For more on this, please refer to https://docs.gradle.org/8.7/userguide/gradle_daemon.html#sec:disabling_the_daemon in the Gradle documentation.
logstash-1  | Daemon will be stopped at the end of the build 
logstash-1  | 
logstash-1  | > Task :compileTestJava
logstash-1  | warning: [options] source value 8 is obsolete and will be removed in a future release
logstash-1  | warning: [options] target value 8 is obsolete and will be removed in a future release
logstash-1  | warning: [options] To suppress warnings about obsolete options, use -Xlint:-options.
logstash-1  | 3 warnings
logstash-1  | 
logstash-1  | BUILD SUCCESSFUL in 4s
logstash-1  | 4 actionable tasks: 2 executed, 2 up-to-date
              /usr/share/plugins/plugin/gradlew clean vendor
logstash-1  | Starting a Gradle Daemon, 1 incompatible and 1 stopped Daemons could not be reused, use --status for details
logstash-1  | 
logstash-1  | > Task :compileJava
logstash-1  | warning: [options] source value 8 is obsolete and will be removed in a future release
logstash-1  | warning: [options] target value 8 is obsolete and will be removed in a future release
logstash-1  | warning: [options] To suppress warnings about obsolete options, use -Xlint:-options.
logstash-1  | 3 warnings
logstash-1  | 
logstash-1  | BUILD SUCCESSFUL in 2s
logstash-1  | 6 actionable tasks: 5 executed, 1 up-to-date
              /usr/share/logstash/vendor/jruby/bin/jruby -I/usr/share/logstash/vendor/jruby/lib/ruby/gems/shared/gems/rspec-core-3.13.6/lib:/usr/share/logstash/vendor/jruby/lib/ruby/gems/shared/gems/rspec-support-3.13.6/lib /usr/share/logstash/vendor/jruby/lib/ruby/gems/shared/gems/rspec-core-3.13.6/exe/rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb
logstash-1  | -XX:+AutoCreateSharedArchive -XX:ConcGCThreads=3 -XX:G1ConcRefinementThreads=10 -XX:GCDrainStackTargetSize=64 -XX:InitialHeapSize=195769920 -XX:MarkStackSize=4194304 -XX:MaxHeapSize=3132318720 -XX:MinHeapSize=6815736 -XX:+PrintCommandLineFlags -XX:ReservedCodeCacheSize=251658240 -XX:+SegmentedCodeCache -XX:SharedArchiveFile=/usr/share/logstash/vendor/jruby/lib/jruby-java21.0.8.jsa -XX:ThreadStackSize=2048 -XX:+UseCompressedOops -XX:+UseG1GC 
logstash-1  | jruby 9.4.13.0 (3.1.4) 2025-06-10 9938a3461f OpenJDK 64-Bit Server VM 21.0.8+9-LTS on 21.0.8+9-LTS +indy +jit [aarch64-linux]
logstash-1  | Sending Logstash logs to null which is now configured via log4j2.properties
logstash-1  | override locale
logstash-1  | Run options: exclude {:integration=>true, :redis=>true, :socket=>true, :performance=>true, :couchdb=>true, :elasticsearch=>true, :elasticsearch_secure=>true, :export_cypher=>true, :windows=>true}
logstash-1  | 
logstash-1  | Randomized with seed 39780
logstash-1  | 
logstash-1  | LogStash::Filters::Date
logstash-1  |   parsing with UNIX
logstash-1  |     "{\"mydate\":\"0\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"%{bad_value}\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"1478207457\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":1000000000}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":1478207457}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"1000000000\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":0}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   LOGSTASH-34 - Default year should be this year
logstash-1  |     "Sun Jun 02 20:38:03"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   failed parses should not cause a failure (LOGSTASH-641)
logstash-1  |     agent(/usr/share/plugins/plugin/spec/filters/date_spec.rb:261:in) runs
logstash-1  |   TAI64N support
logstash-1  |     "{\"t\":\"4000000050d506482dbdf024\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"t\":\"@4000000050d506482dbdf024\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   success to parse should apply on_success config(add_tag,add_field...)
logstash-1  |     "{\"thedate\":\"2013/04/21\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   failing to parse should apply tag_on_failure
logstash-1  |     "{\"thedate\":\"2013/Apr/21\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   accept match config option with hash value (LOGSTASH-735)
logstash-1  |     "{\"mydate\":\"2001-09-09T01:46:40.000Z\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   Default year handling when parsing with english fallback parser
logstash-1  |     LOGSTASH-34 - Default year should be this year
logstash-1  |       "Sun Jun 02 20:38:03"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |         processes events as specified
logstash-1  |     fill next year if january events arrive in december
logstash-1  |       "Jan 01 01:00:00"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |         processes events as specified
logstash-1  |     fill last year if december events arrive in january
logstash-1  |       "Dec 31 23:59:00"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |         processes events as specified
logstash-1  |   parsing IS08601 with timezone from event
logstash-1  |     "{\"mydate\":\"2007-12-09T01:00:00\",\"mytz\":\"America/Car..."
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2007-12-09T10:00:00\",\"mytz\":\"America/Car..."
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2016-05-01T08:18:18.123\",\"mytz\":\"America..."
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   cancelled events
logstash-1  |     cancelled events list
logstash-1  |       ignores and return ignored cancelled
logstash-1  |     single cancelled event
logstash-1  |       ignores and return cancelled
logstash-1  |   Supporting locale+country in BCP47
logstash-1  |     "14 juillet 1789"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   don't fail on next years DST switchover in CET
logstash-1  |     "2016 Mar 26 02:00:37"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   parsing microsecond-precise times with UNIX (#213)
logstash-1  |     "{\"mydate\":1350414944.123456}"
logstash-1  | /usr/share/plugins/plugin/spec/filters/date_spec.rb:172: warning: BigDecimal.new is deprecated; use BigDecimal() method instead.
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":1.350414944123456E9}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"1350414944.123456\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"%{bad_value}\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   when giving an invalid match config
logstash-1  |     raises a configuration error
logstash-1  |   parsing with timezone parameter
logstash-1  |     "{\"mydate\":\"2013 Nov 24 01:29:01\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2013 Jun 24 01:29:01\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   failing to parse should not apply on_success config(add_tag,add_field...)
logstash-1  |     "{\"thedate\":\"2013/Apr/21\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   Default year handling when parsing with timezone from event
logstash-1  |     fill last year if december events arrive in january
logstash-1  |       "{\"message\":\"Dec 31 23:59:00\",\"mytz\":\"UTC\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |         processes events as specified
logstash-1  |     fill next year if january events arrive in december
logstash-1  |       "{\"message\":\"Jan 01 01:00:00\",\"mytz\":\"UTC\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |         processes events as specified
logstash-1  |     LOGSTASH-34 - Default year should be this year
logstash-1  |       "{\"message\":\"Sun Jun 02 20:38:03\",\"mytz\":\"UTC\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |         processes events as specified
logstash-1  |     do fail on 2016 DST switchover in CET
logstash-1  |       "Mar 27 02:00:01"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |         processes events as specified
logstash-1  |       "Mar 27 03:00:01"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |         processes events as specified
logstash-1  |       "Mar 27 01:59:59"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |         processes events as specified
logstash-1  |   support deep nested field access
logstash-1  |     "{\"data\":{\"deep\":\"2013-01-01T00:00:00.000Z\"}}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   parsing with java SimpleDateFormat syntax
logstash-1  |     "{\"mydate\":\"Nov 24 01:29:01 -0800\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   Supporting locale only
logstash-1  |     "14 juillet 1789"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   metric counters
logstash-1  |     when date parses a date correctly
logstash-1  |       increases the matches counter
logstash-1  |     when date parses a date correctly
logstash-1  |       increases the matches counter
logstash-1  |   fill next year if january events arrive in december
logstash-1  |     "Jan 01 01:00:00"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   http dates
logstash-1  |     "{\"timestamp\":\"25/Mar/2013:20:33:56 +0000\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   fill last year if december events arrive in january
logstash-1  |     "Dec 31 23:59:00"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   Supporting locale+country in POSIX (internally replace '_' by '-')
logstash-1  |     "14 juillet 1789"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   parsing with UNIX and UNIX_MS
logstash-1  |     "{\"mydate\":1.478207457456E9}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":1000000000123}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"0\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":1478207457}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":0}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"1000000000\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"1478207457\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":1000000000}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"1478207457.456\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"1000000000123\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   parsing with UNIX_MS
logstash-1  |     "{\"mydate\":1000000000123}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"1000000000123\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":456}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"456\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":0}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"0\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   Support fallback to english for non-english default locale
logstash-1  |     "01 September 2014"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   parsing with ISO8601
logstash-1  |     "{\"mydate\":\"2004-07-04T12:27:27-00:00\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2001-11-06 20:45:45.123-0000\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2001-09-05T16:36:36+0000\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2001-11-06T20:45:45-0000\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2004-07-04T12:27:27.123-04:00\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2001-12-07T23:54:54Z\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2001-11-06T20:45:45.123-0000\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2010-05-03T08:18:18.123+00:00\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2001-12-07T23:54:54.123Z\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"1974-03-02T04:09:09-0800\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2001-11-06 20:45:45.123\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2001-12-07 23:54:54,123Z\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2001-12-07T23:54:54,123Z\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2001-09-05T16:36:36.123+0700\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2001-12-07 23:54:54.123Z\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2001-01-01T00:00:00-0800\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2010-05-03T08:18:18+00:00\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2001-11-06 20:45:45,123\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   failing to parse should not throw an exception
logstash-1  |     "{\"thedate\":\"2013/Apr/21\"}"
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |   parsing with timezone from event
logstash-1  |     "{\"mydate\":\"2013 Jun 24 01:29:01\",\"mytz\":\"America/Lo..."
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  |     "{\"mydate\":\"2013 Nov 24 01:29:01\",\"mytz\":\"America/Lo..."
logstash-1  | Accessing filtered_events
logstash-1  | Accessing filtered_events
logstash-1  |       processes events as specified
logstash-1  | 
logstash-1  | Finished in 16.62 seconds (files took 2.01 seconds to load)
logstash-1  | 85 examples, 0 failures
logstash-1  | 
logstash-1  | Randomized with seed 39780
logstash-1  | 
logstash-1 exited with code 0                                                                                                                                                                             
Aborting on container exit...
 Container ci-logstash-1  Stopping
 Container ci-logstash-1  Stopped
```

donoghuc reviewed (approved):
Thanks for the manual testing description ❤️
